### PR TITLE
Provide a suggestion when trying to destructure a `Vec` as a slice

### LINKED ISSUE #87017: When trying to destructure a `Vec` as a slice provide a suggestion
We currently emit the following:

```
error[E0529]: expected an array or slice, found `Vec<types::GenericBound>`
   --> src/librustdoc/clean/mod.rs:593:28
    |
593 |                     if let [] | [GenericBound::TraitBound(_, hir::TraitBoundModifier::Maybe)] = &bounds {
    |                            ^^ pattern cannot match with input type `Vec<types::GenericBound>`

error[E0529]: expected an array or slice, found `Vec<types::GenericBound>`
   --> src/librustdoc/clean/mod.rs:593:33
    |
593 |                     if let [] | [GenericBound::TraitBound(_, hir::TraitBoundModifier::Maybe)] = &bounds {
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ pattern cannot match with input type `Vec<types::GenericBound>`
```

Ideally we should mention the appropriate solution, to use `&bounds[..]` instead.

### PR DESCRIPTION
**Pull request by FabianWolff**
Fixes #87017.

r? @estebank

### CODE REVIEW

File: compiler/rustc_typeck/src/check/pat.rs [hunk 1]
```diff
@@ -1894,6 +1894,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                        err.span_suggestion(
+                            span,
+                            "consider slicing here",
+                            format!("{}[..]", snippet),
+                            Applicability::MachineApplicable,
+                        );
+                    }
+                }
+            }
```
Comment (camsteffen) [inline compiler/rustc_typeck/src/check/pat.rs:1914]: Can you use `AutoDeref::new().find()` to see if any type derefs to slice?

Comment (FabianWolff) [reply compiler/rustc_typeck/src/check/pat.rs:1911]: Fixed, thanks for the suggestion!

Comment (camsteffen) [reply compiler/rustc_typeck/src/check/pat.rs:1911]: Nice! I wasn't sure if that would work. You can replace `find().is_some()` with `any()`.

**Comment by estebank** [comment]
r=me after applying the change (and re-running `--bless` to generate the `.fixed` file)

File: src/test/ui/suggestions/pattern-slice-vec.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Regression test for #87017.
```
[the remark this answers is not in the document]
Comment (FabianWolff) [reply src/test/ui/suggestions/pattern-slice-vec.rs:1]: Fixed. I've also added your `if let [] = foo()` example to the test case.

**Comment by FabianWolff** [comment]
> r=me after applying the change (and re-running `--bless` to generate the `.fixed` file)

This is ready to be merged. Thanks for the review @estebank!

**Pull request merged**